adding articles under /topics collection (#2320)
layer5io/layer5: src/collections/topics/service-mesh-architectures-and-components/index.mdx

---
title: "Service Mesh Architecture and Components"
thumbnail: ./service-mesh.svg
category: Service mesh
tags:
 - Service mesh
 - Network Planes
featured: false
published: true
---

import { Link } from "gatsby";
import { TopicsWrapper } from "../Topics.style.js";
import Planes from "./epsm_0101.png";
import Topology from "./epsm_0102.png";
import Architecture from "./epsm_0103.png";
import Meshery from "./epsm_0112.png";

<TopicsWrapper>
  <div className="intro">
    <p>Learn more about service mesh fundamentals in

>       <Link className="blog" to="/learn/books">The Enterprise Path to Service Mesh Archictures (2nd Edition)</Link> -  free book and excellent resource which addresses how to evaluate your organization’s readiness, provides factors to consider when building new applications and converting existing applications to best take advantage of a service mesh, and offers insight on deployment architectures used to get you there.</p>

Reviewer: ```suggestion
      <Link className="blog" to="/learn/books">The Enterprise Path to Service Mesh Archictures (2nd Edition)</Link> -  an excellent resource which addresses how to evaluate your organization’s readiness, provides factors to consider when building new applications and converting existing applications to best take advantage of a service mesh, and offers insight on deployment architectures used to get you there.</p>
```
(Optional) the `free book` part bothered me a bit.

</div>

<p>Envoy is an example of a popular proxy used in service mesh data planes. It is also often found deployed more simply standalone as a load balancer or ingress gateway. The proxies used in service mesh data planes are highly intelligent and may incorporate any number of protocol-specific filters to manipulate network packets (including application level data). With technology advances like WebAssembly, extending data plane capabilities means that service meshes are capable of injecting new logic into requests while simultaneously handling high traffic load.
</p>

<p>

>   A service mesh control plane is called for when the number of proxies becomes unwieldy or when a single point of visibility and control is required. Control planes provide policy and configuration for services in the mesh, taking a set of isolated, stateless proxies and turning them into a service mesh. Control planes do not directly touch any network packets in the mesh; they operate out-of-band. Control planes typically have a command-line interface (CLI) and user interface with which to interact, each of which provides access to a centralized API for holistically controlling proxy behavior. You can automate changes to the control plane configuration through its APIs (e.g., by a continuous integration/continuous deployment pipeline), where, in practice, configuration is most often version controlled and updated.

Reviewer: ```suggestion
  A service mesh control plane is called for when the number of proxies becomes unwieldy or when a single point of visibility and control is required. Control planes provide policy and configuration for services in the mesh, taking a set of isolated, stateless proxies and turning them into a service mesh. Control planes do not directly touch any network packets in the mesh; they operate out-of-band. Control planes typically have a command-line interface (CLI) and a user interface to interact with, each of which provides access to a centralized API for holistically controlling proxy behavior. You can automate changes to the control plane configuration through its APIs (e.g., by a continuous integration/continuous deployment pipeline), where, in practice, configuration is most often version controlled and updated.
```

</p>

<div class="fact">
  Proxies are generally considered stateless, but this is a thought-provoking concept. In the way in which proxies are generally informed by the control plane of the presence of services, mesh topology updates, traffic and authorization policy, and so on, proxies cache the state of the mesh but aren’t regarded as the source of truth for the state of the mesh.
</div>

<p>

>   Reflecting on Linkerd (pronounced “linker-dee”) and Istio (pronounced “Ist-tee-oh”) as two popular, open source service meshes, we find examples of how the data and control planes are packaged and deployed. In terms of packaging, Linkerdv1 contains both its proxying components (linkerd) and its control plane (namerd) packaged together simply as “Linkerd,” and Istio brings a collection of control plane components (Galley, Pilot, and Citadel) to pair by default with Envoy (a data plane) packaged together as “Istio.” Envoy is often labeled a service mesh, inappropriately so, because it takes packaging with a control plane to form a service mesh. 

Reviewer: ```suggestion
  Reflecting on Linkerd (pronounced “linker-dee”) and Istio (pronounced “Ist-tee-oh”) as two popular, open source service meshes, we find examples of how the data and control planes are packaged and deployed. In terms of packaging, Linkerdv1 contains both its proxying components (linkerd) and its control plane (namerd) packaged together simply as “Linkerd”, and Istio brings a collection of control plane components (Galley, Pilot, and Citadel) to pair by default with Envoy (a data plane), all packaged together as “Istio.” Envoy is often labeled a service mesh, inappropriately so, because it takes packaging with a control plane to form a service mesh. 
```
